### ISSUE TITLE
Return types aren't checked on promises

### ISSUE THREAD

**Issue by Yannic** [google/closure-compiler#2582]
[Repo CHECK_TYPES](https://yabopu.sh/41Y2epHxFn3lGL08)  
[Repo CHECK_TYPES_NEW_INFERENCE](https://yabopu.sh/41Y2epHxFlMmAW4J)

The following code is not checked correctly.  
OTI catches inconsistent return types only with empty `then` or `catch` methods while NTI does not catch any error at all.

```javascript
/**
 * @return {!Promise<boolean>}
 */
function foo() {
  return Promise.resolve(true);
}

/**
 * @return {!Promise<boolean>}
 */
function foo2() {
  // No warning expected.
  return foo();
}

/**
 * @return {!Promise<boolean>}
 */
function bar() {
  // FOUND (OTI):
  //     input0:33: WARNING - inconsistent return type
  //     found   : Promise<undefined>
  //     required: Promise<boolean>
  //       return foo().catch(() => {});
  //              ^^^^^^^^^^^^^^^^^^^^^
  // FOUND (NTI):
  //     No warning
  // EXPECTED:
  //     input0:33: WARNING - inconsistent return type
  //     found   : Promise<boolean|undefined>
  //     required: Promise<boolean>

  return foo().catch(() => {});
}

/**
 * @return {!Promise<boolean>}
 */
function foobar() {
  // FOUND:
  //     No warning
  // EXPECTED:
  //     input0:49: WARNING - inconsistent return type
  //     found   : Promise<undefined>
  //     required: Promise<boolean>
  //       return foo().then(() => {
  //              ^^^^^^^^^^^^^^^^^^^

  return foo().then(() => {
    return Promise.resolve('not a boolean');
  });
}
```

**Comment by dimvar** [comment]
This happens because NTI does not yet handle the type transformation language. We are currently working on fixing that. Should be done sometime next week.

**Comment by dimvar** [comment]
Closing this. NTI now handles the TTL. The remaining issues here are because the callbacks are not type annotated, and inference isn't as good in this case.
